### PR TITLE
fix: potential `async_hooks` crash in `NotifyWindowRestore` on Windows

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/40573.

Fixes an issue where an async_hooks crash could occur on Windows only if the user is listening to the `restore` event after minimizing a maximized window and an error occurs in the `restore` callback. This was happening after https://github.com/electron/electron/commit/6b97beb4897b47c1bad698c45bc01fbdae162ceb, which created the first instance in our entire codebase where two `Emit` calls were called sequentially. As a result, given they were in the same callbackscope, when `node::RunTimers` ran and an error was occurring in the previous handler, async_hooks would blow up.

This fixes that error by putting the first `Emit` in its own `CallbackScope`.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Fixed a potential `async_hooks` crash when listening for the `restore` event on Windows after minimizing a maximized BrowserWindow.

### CODE REVIEW

**Review by VerteDinde — APPROVED** [review]

**Review by zcbenz — COMMENTED** [review]
Does it fix the problem if we wrap every `Emit` with a `node::CallbackScope`? i.e. Put the `node::CallbackScope` in `CallMethodWithArgs` at `event_emitter_caller.cc`.

Sequential emits is rare but it seems impossible to patch all corner cases, for example, the `drag-ended` and `drop` events of `Tray` emit in one tick:
https://github.com/electron/electron/blob/main/shell/browser/ui/tray_icon_cocoa.mm#L303-L309

So I would prefer a generic solution then patching each edge case.

**Review by zcbenz — APPROVED** [review]

**Comment by zcbenz** [comment]
The failing linux-x64-testing-asan-tests does not seem to be flaky.

**Comment by codebytere** [comment]
Unrelated failure on macOS:

> Error: Could not get code signature for running application

File: shell/browser/api/electron_api_auto_updater.cc [hunk 1]
```diff
@@ -32,13 +32,26 @@ void AutoUpdater::OnError(const std::string& message) {
...
+  // put this in its own CallbackScope and delegate to Node.js'
+  // specialized handling for Error objects.
   if (GetWrapper(isolate).ToLocal(&wrapper)) {
     auto error = v8::Exception::Error(gin::StringToV8(isolate, message));
-    gin_helper::EmitEvent(
-        isolate, wrapper, "error",
-        error->ToObject(isolate->GetCurrentContext()).ToLocalChecked(),
-        // Message is also emitted to keep compatibility with old code.
-        message);
+    std::vector<v8::Local<v8::Value>> args = {
```
Comment (codebytere) [inline shell/browser/api/electron_api_auto_updater.cc:41]: `autoUpdater` is then only place in the codebase we use gin to emit a proper `Error` object. By using `gin::EmitEvent` to emit the Error object (thus adding a new callback scope ther), and because Errors are handled specially in Node.js, we would otherwise in triggering the uncaught exception handler to bypass the try/catch in our autoUpdate specs. Either we add CallbackScopes on a case-by-case basis (as i initially tried in this PR) or we account for this edge case. I've chosen to account for this edge case.

**Review by zcbenz — APPROVED** [review]
